Expand non-transparent macros after Typer

nicolasstucki opened the pull request:
The technical change 
* Inline `transparent inline` definitions in typer
* Inline other `inline` after typer.

Sematic changes
* `inline given`
  * `transparent inline given f: T = ...`: does the same as all inline definition did before
    * `transparent inline given f: T = (...): T` can be used to emulate the previous version `inline given`
  * `inline given f: T = ...`: if an error is emitted during the inlining then the error is reported to the user.
* `scala.compiletime.testing.{typeChecks, typeCheckErrors}` must be used directly or inside a `transparent inline`
* Add missing type parameter bound checks for non-transparent inlines (see #10552). Transparent inlines still have this issue.

deusaquilus commented:
This looks really scary to me. Is there a build/branch I can try it with somewhere?

nicolasstucki commented:
Not yet. 

An extra note is that the current behavior of non-transparent/transparent macros will not be lost. It will be possible to emulate the current non-transparent macros using a transparent one. I will add documentation in one of the `.md` at some point.

deusaquilus commented:
Will that be necessary for whitebox macros only or also for blackbox? It would be really nasty if all variables holding quill quotes would have to be `transparent inline def`.

nicolasstucki commented:
TODOs
- [ ] Fix Scalatest
- [ ] Fix izumiReflect
- [ ] Fix sbt-dotty tests
- [ ] Update docs

deusaquilus commented:
Should I be testing against this change now?

nicolasstucki commented:
No, changes will be minimal

nicolasstucki commented:
test performance please

nicolasstucki commented:
I am not sure how to debug izumiReflect as it just tells me that the rest failed but I don't see any output.

```
sbt:izumi-reflect-root> test
[error] Failed tests:
[error]         izumi.reflect.test.LightTypeTagTest
[error] (izumi-reflect / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 123 s (02:03), completed Feb 2, 2021 11:19:27 AM
```

 @neko-kai do you know how I could see what happened in that test?

nicolasstucki commented:
@odersky @liufengyun this PR is ready to be reviewed. I'm still working on figuring out what is failing in `izumi-reflect`.

neko-kai commented:
@nicolasstucki 
This looks specific to the way community build outputs test results - I mean, izumi-reflect doesn't do anything special, the tests are vanilla `scalatest`. I've tried to replicate and on upstream repository errors are shown correctly:
```
[info] - should support self-intersection (X with X) *** FAILED *** (25 milliseconds)
[info]   t.=:=(expected) was true String =!= String (TagAssertions.scala:51)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:476)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:421)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1344)
...
[info] Run completed in 643 milliseconds.
[info] Total number of tests run: 22
[info] Suites: completed 4, aborted 0
[info] Tests: succeeded 21, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error]         izumi.reflect.test.LightTypeTagTest
[error] (izumi-reflect / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 92 s (01:32), completed Feb 2, 2021 7:04:50 PM
```
to reproduce execute
```
git clone https://github.com/zio/izumi-reflect.git --branch fail-test
cd izumi-reflect
sbt test
```

nicolasstucki commented:
Found the issue, we set `logLevel in Global := Level.Error` in the community build.

liufengyun commented on compiler/src/dotty/tools/dotc/transform/Inlining.scala:
```diff
@@ -87,4 +105,19 @@ class Inlining extends MacroTransform {
 object Inlining {
   val name: String = "inlining"
+
+  /** Check that `vdef.rhs` can be right hand-side or argument to `inline` value or parameter. */
```
```suggestion
  /** Check that `vdef.rhs` can be right hand-side of an `inline` value definition. */
```

liufengyun commented on compiler/src/dotty/tools/dotc/transform/PostTyper.scala:
```diff
@@ -257,12 +257,16 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
       try tree match {
         case tree: Ident if !tree.isType =>
+          if tree.symbol.is(Inline) then
+            ctx.compilationUnit.needsInlining = true
```
Check `ctx.owner` is not `inline`?

nicolasstucki replied:
Yes. Will add the check

liufengyun commented on compiler/src/dotty/tools/dotc/transform/PostTyper.scala:
```diff
@@ -326,6 +335,8 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
           }
         case tree: ValDef =>
+          if tree.symbol.is(Inline, butNot = Param) then
+            ctx.compilationUnit.needsInlining = true
```
It's not clear to me why we need this.

nicolasstucki replied:
Now `Inlining` checks that the rhs of inline vals are constants. Technically we could do this in a later phase.

liufengyun commented on compiler/src/dotty/tools/dotc/CompilationUnit.scala:
```diff
@@ -38,15 +38,26 @@ class CompilationUnit protected (val source: SourceFile) {
+   *  The information is used in phase `PickleQuotes` in order to avoid traversing trees that need no transformations.
+   */
+  var needsQuotePickling: Boolean = false
+
```
Do we need both `needsStaging` and `needsQuotePickling`?

nicolasstucki replied:
There are some cases where we only need one of them. Each causes a full tree traversal as these are `MacroTransform` phases.

liufengyun commented on compiler/src/dotty/tools/dotc/transform/Inlining.scala:
```diff
@@ -33,11 +33,22 @@ import scala.annotation.constructorOnly
+      catch case _: CompilationUnit.SuspendException => ()
+
+  override def runOn(units: List[CompilationUnit])(using Context): List[CompilationUnit] =
+    val newUnits = super.runOn(units).filterNot(_.suspended)
```
Is it `suspendedAtInliningPhase` instead of `suspended`? Or do we really need `suspendedAtInliningPhase`?

```suggestion
    val newUnits = super.runOn(units).filterNot(_. suspendedAtInliningPhase)
```

nicolasstucki replied:
No, it needs to be `.suspended` as we do not want to transform compilation units that were suspended in typer.

`suspendedAtInliningPhase` is only needed to let `sbt.ExtractDependencies` know that it has already registered the classes of this compilation unit.

liufengyun commented on docs/docs/reference/metaprogramming/inline.md:
````diff
@@ -284,6 +284,21 @@ transparent inline def zero: Int = 0
 ```
 
+### Transparent vs. non-transparent inline
+As we already discussed, transparent inline methods will provide a more precise.
````
```suggestion
As we already discussed, transparent inline methods may influence type checking.
```

liufengyun commented on docs/docs/reference/metaprogramming/inline.md:
```diff
@@ -284,6 +284,21 @@ transparent inline def zero: Int = 0
 
+### Transparent vs. non-transparent inline
+As we already discussed, transparent inline methods will provide a more precise.
+Technically this implies that transparent inline methods must be expanded while typing the program. Other inline methods are inlined later after the program is fully typed.
```
```suggestion
Technically this implies that transparent inline methods must be expanded during type checking of the program. Other inline methods are inlined later after the program is fully typed.
```

liufengyun commented on docs/docs/reference/metaprogramming/inline.md:
```diff
@@ -284,6 +284,21 @@ transparent inline def zero: Int = 0
+As we already discussed, transparent inline methods will provide a more precise.
+Technically this implies that transparent inline methods must be expanded while typing the program. Other inline methods are inlined later after the program is fully typed.
+
+For example the following two functions will be typed the same way but will be inlined at different times.
```
```suggestion
For example, the following two functions will be typed the same way but will be inlined at different times.
```

liufengyun commented on docs/docs/reference/metaprogramming/inline.md:
```diff
@@ -284,6 +284,21 @@ transparent inline def zero: Int = 0
+A noteworthy difference is the behavior of `transparent inline given`.
+If there is an error reported when inlining that definition, it will be considered as an implicit search mismatch and the search will continue.
+A `transparent inline given` can add a type ascription in its RHS (as in `f2` from the previous example) to avoid the precise type but keep the search behavior.
+On the other hand `inline given` be taken as the implicit and then after typing is done the code is inlined and any error will be emitted as usual.
```
```suggestion
On the other hand, `inline given` be taken as the implicit and then inlined after typing. Any error will be emitted as usual.
```

liufengyun commented on docs/docs/reference/metaprogramming/inline.md:
````diff
@@ -313,6 +328,8 @@ below:
    | This location is in code that was inlined at ...
 ```
 
+In a transparent inline, an `inline if` will force the inlining of any inline definition in its condition.
````
Can it still expand after typer?

nicolasstucki replied:
Yes

liufengyun commented on library/src/scala/compiletime/testing/package.scala:
```diff
@@ -2,14 +2,16 @@ package scala.compiletime
-inline def typeChecks(inline code: String): Boolean =
+transparent inline def typeChecks(inline code: String): Boolean =
   // implemented in package dotty.tools.dotc.typer.Inliner.Intrinsics
   error("Compiler bug: `typeChecks` was not checked by the compiler")
```
If we handle the intrinsics in `Inlining` phase, it seems we don't need `transparent`. Same for below.

nicolasstucki replied:
`typeChecks` has an interaction with the typer state the might not be possible after typer.

https://github.com/lampepfl/dotty/blob/8443c16b7d5f965c23418d5b13aaa195116842a1/compiler/src/dotty/tools/dotc/typer/Inliner.scala#L294

liufengyun replied:
If we have the correct context that contains the right scope with imports, it seems to be possible.

nicolasstucki replied:
We do not as the only typer we have when inlining in the InlineTyper and that one starts when expanding the inline call.

nicolasstucki replied:
Maybe it is the import info that is not set correctly in `Inlining`

liufengyun commented on library/src/scala/runtime/stdLibPatches/Predef.scala:
```diff
@@ -3,10 +3,10 @@ package scala.runtime.stdLibPatches
 
-  inline def assert(inline assertion: Boolean, inline message: => Any): Unit =
+  transparent inline def assert(inline assertion: Boolean, inline message: => Any): Unit =
     if !assertion then scala.runtime.Scala3RunTime.assertFailed(message)
```
Seems it can be inlined later? The same for below.

nicolasstucki replied:
This one has a special interaction with null safety that requires it to be inlined in typer. It would be much better to inline after typer.

liufengyun reviewed (commented):
Great job 👍 

Overall, LGTM. I've some comments about CB changes.

It's not obvious to be why the macros need to be `transparent` in the CB projects below:

- scalatest, it seems we can do without `transparent`
- `izumi-reflect`: no need to be transparent?
- `shapeless`: `mkProductInstances` can be non-transparent?
- `utest`:  remove `transparent`?

nicolasstucki commented on compiler/src/dotty/tools/dotc/CompilationUnit.scala:
```diff
@@ -116,10 +131,13 @@ object CompilationUnit {
-        needsStaging = true
+        containsQuote = true
+      if tree.symbol.is(Flags.Inline) then
+        containsInline = true
```
This only happens when unpickling a quote and we do not allow inline definitions in quotes. Therefore it should be ok.

nicolasstucki commented on compiler/src/dotty/tools/dotc/core/Mode.scala:
```diff
@@ -122,4 +122,9 @@ object Mode {
+  /** We are typing the body of the condition of an `inline if` or the scrutinee of an `inline match`
+   *  This mode forces expansion of inline calls in those positions even during typing.
+   */
+  val ForceInline: Mode = newMode(29, "ForceInline")
```
Yes. Any inline if that is found in the `Inlining` phase is reduced there. This happens with all `inline if` in non-transparent inline methods.

nicolasstucki commented:
@liufengyun 

> * scalatest, it seems we can do without transparent

Those are needed for `typeChecks`, if one can be removed it might indicate that the tests are not exahstive.

> * izumi-reflect: no need to be transparent?

There are some transparents that do not need to be there. But if I remove them I fix #10552 which then leads to tests that do not compile. I added a comment and will let them have a look at those tests.

> * shapeless: mkProductInstances can be non-transparent?

I'm not sure why it is needed in the other ones. It looks like another case where match types might not be normalized enough. Like in #11236 and #11247.

> * utest: remove transparent?

Removed all but one.
It is needed for `typeChecks`, if one can be removed it might indicate that the tests are not exahstive.

odersky commented on compiler/src/dotty/tools/dotc/typer/Inliner.scala:
```diff
@@ -1335,6 +1357,16 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
+
+    private def inlineIfIsNestedInlineCall(tree: Tree)(using Context): Tree =
+      if
+        !suppressInline &&
```
Cosmetic: I think it reads better of the `&&` comes first on each subsequent line

nicolasstucki replied:
Done

odersky commented on compiler/src/dotty/tools/dotc/typer/Inliner.scala:
```diff
@@ -1335,6 +1357,16 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
     override def suppressInline(using Context) =
       ctx.reporter.errorCount > initialErrorCount || super.suppressInline
+
+    private def inlineIfIsNestedInlineCall(tree: Tree)(using Context): Tree =
```
This looks like a change in behavior previously we just returned the call and Typer would do the re-inlining. Why is that now done here? Typer already contains the logic to do repeated inline calls in apply. Should there be a similar call in Inliner? I.e. a single toplevel call, instead of 4 in different specific places?

nicolasstucki replied:
The issue is that when we are inlining in the `Inlining` phase there is no normal typer to inline those nested calls.

nicolasstucki replied:
I will try to see if it is possible to have a single place to inline nested calls.

nicolasstucki replied:
Refactored the code in 53d69cdeb54111ce5b23c6a70ccb9b87016496b9

nicolasstucki replied:
@odersky could you review that last commit before merging?

nicolasstucki commented on bench-run/src/main/scala/dotty/tools/benchmarks/tuples/TupleOps.scala:
```diff
@@ -50,7 +50,7 @@ class TupleOps {
     if (tuple.size <= 1) tuple
     else {
-      val (tuple1, tuple2) = tuple.splitAt(tuple.size / 2)
+      val (tuple1, tuple2) = tuple.splitAt(tuple.size / 2): (Tuple, Tuple)// TODO remove ascriptions (issue with type variable constraints)
```
Those are imports of extension methods that are in the correct path to not require the given imports. We needed those a long time ago, but users have kept copying the old version anyway.

nicolasstucki replied:
I will open an issue for it. It looks like it is something related to #11247.

<details>

```scala
53 |      val (tuple1, tuple2) = tuple.splitAt(tuple.size / 2)
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |Found:    (scala.Tuple.Take[Tuple, (?1 : Int)], scala.Tuple.Drop[Tuple, (?1 : Int)])
   |Required: (
   |  Int match {
   |    case (0 : Int) => EmptyTuple
   |    case 
   |      [n1 <: Int] =>> 
   |        scala.runtime.MatchCase[compiletime.S[n1], 
   |          Tuple match {
   |            case EmptyTuple => EmptyTuple
   |            case 
   |              [x, xs <: Tuple] =>> 
   |                scala.runtime.MatchCase[x *: xs, x *: scala.Tuple.Take[xs, n1]]
   |          } <: Tuple
   |        ]
   |  } <: Tuple
   |, 
   |  Int match {
   |    case (0 : Int) => Tuple
   |    case 
   |      [n1 <: Int] =>> 
   |        scala.runtime.MatchCase[compiletime.S[n1²], 
   |          Tuple match {
   |            case EmptyTuple => EmptyTuple
   |            case 
   |              [x, xs <: Tuple] =>> 
   |                scala.runtime.MatchCase[x² *: xs², scala.Tuple.Drop[xs², n1²]]
   |          } <: Tuple
   |        ]
   |  } <: Tuple
   |)
   |
   |where:    ?1  is an unknown value of type Int
   |          n1  is a type variable with constraint <: Int
   |          n1² is a type variable with constraint <: Int
   |          x   is a type variable with constraint 
   |          xs  is a type variable with constraint <: Tuple
   |          xs² is a type variable with constraint <: Tuple
   |          x²  is a type variable with constraint 
```

</details>

liufengyun commented:
@nicolasstucki  Maybe we can create a separate issue for `typeChecks` 

https://github.com/lampepfl/dotty/blob/8b1188b7f5a9f588ed94acfc46896089d0c71930/library/src/scala/compiletime/testing/package.scala#L12-L14

and see how we can avoid adding `transparent` and avoid making related macros transparent in ScalaTest.

Conceptually, it should work, as macros that expanded later also depend on the correct context for the reflect API `Implicits.search`.

nicolasstucki commented:
test performance please

odersky commented:
test performance please

odersky commented on compiler/src/dotty/tools/dotc/transform/PostTyper.scala:
```diff
@@ -272,6 +276,8 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
             checkNoConstructorProxy(tree)
             transformSelect(tree, Nil)
         case tree: Apply =>
+          if tree.symbol.is(Inline) && !Inliner.inInlineMethod then
```
Why is this needed here? Don't we get to an Ident or Select eventually, which sets `needsInlining`. Also, why do we handle `Apply` here but not `TypeApply`. If this is in fact needed for some reason it would be good to add a comment.

nicolasstucki replied:
It is redundant. I removed it.

odersky commented on compiler/src/dotty/tools/dotc/transform/PostTyper.scala:
```diff
@@ -329,6 +338,8 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
             )
           }
         case tree: ValDef =>
+          if tree.symbol.is(Inline, butNot = Param) then
```
Would this be triggered by
```scala
inline val x = 2
```
? Those things are extremely common, so we should avoid running the inliner just for these.

nicolasstucki replied:
Yes, then we would need to check that the RHS of an `inline val` is a constant after the `Inlining` phase. I try to fit it into another phase.

odersky replied:
That would cause many more traversals, so yes, good to do this somewhere else.

odersky commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -3407,13 +3406,13 @@ class Typer extends Namer
           errorTree(tree, i"Deferred inline ${meth.showLocated} cannot be invoked")
-        else if (Inliner.isInlineable(tree) && !suppressInline && StagingContext.level == 0) {
+        else if Inliner.needsInlining(tree)
+        then
```
No need to put `then` on its own line here.

odersky commented on compiler/src/dotty/tools/dotc/typer/Inliner.scala:
```diff
@@ -1235,8 +1254,14 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
       super.ensureAccessible(tpe, superAccess, pos)
     }
 
+    override def typed(tree: untpd.Tree, pt: Type = WildcardType)(using Context): Tree =
```
In fact, it might be better to do this test only in typedApply and typedTypeApply. Concentrating everything in `typed` is more elegant but might have performance implications. I am also nervous about making calls to `typed` more megamorphic.

nicolasstucki replied:
I moved the logic back to `typedApply`, `typedTypeApply`, `typedIdent` and `typedSelect`. I kept the rest of the refactoring which avoids the duplication of the code that is now in `needsInlining`.

odersky commented on compiler/src/dotty/tools/dotc/typer/Inliner.scala:
```diff
@@ -1252,7 +1277,7 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
 
     override def typedIf(tree: untpd.If, pt: Type)(using Context): Tree =
-      typed(tree.cond, defn.BooleanType) match {
+      typed(tree.cond, defn.BooleanType)(using ctx.addMode(Mode.ForceInline)) match {
```
`ForceInline` should only be added if `tree.isInline`.

nicolasstucki replied:
Changed

odersky reviewed (changes requested):
Can we try without overriding typer and test performance again?

nicolasstucki commented:
test performance with #quotes please

nicolasstucki commented:
@liufengyun could you have a look at the new commits?

liufengyun reviewed (approved):
LGTM

deusaquilus commented:
It seems like this is working for me out of the box! I don't need to modify anything.
I do have one question though. If the typing phase is now before blackbox macros, how come the Quill parser is running after a typing mistake.

Say I make a typing mistake on purpose:
```scala
    inline def q = quote {
      query[SanePerson] // hello
    }
    inline def name(p: SanePerson) = p.name
    inline def qq = quote {
      q.map(p => name(p.name)) // purposely making a typing mistake
    }
```

Why does the macro even run if there was an error on a previous phase?

![image](https://user-images.githubusercontent.com/1369480/107903650-cf6ead00-6f17-11eb-84a6-dc6005a61d8b.png)
